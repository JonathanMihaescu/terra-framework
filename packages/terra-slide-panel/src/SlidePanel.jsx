import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import classNamesBind from 'classnames/bind';
import ThemeContext from 'terra-theme-context';
import VisuallyHiddenText from 'terra-visually-hidden-text';
import { injectIntl } from 'react-intl';
import styles from './SlidePanel.module.scss';
import SlidePanelUtils from './_SlidePanelUtils';

const cx = classNamesBind.bind(styles);

const SlidePanelPositions = {
  START: 'start',
  END: 'end',
};

const propTypes = {
  /**
   * String that labels the Panel content area for screen readers.
   */
  panelAriaLabel: PropTypes.string,

  /**
   * String that labels the Main content area for screen readers.
   */
  mainAriaLabel: PropTypes.string,

  /**
   * An ID or space-separated string of IDs that describe the Main content area for screen readers. For MacOS VoiceOver only.
   * Example: "patient-details" or "patient-details-1 patient-details-2 patient-details-3" are valid.
   */
  mainAriaDescribedBy: PropTypes.string,

  /**
   * Whether the mainAriaDescribedBy should replace the default or be appended to it. For   MacOS VoiceOver only.
   */
  replaceMainAriaDescribedBy: PropTypes.bool,

  /**
   * The component to display in the main content area.
   */
  mainContent: PropTypes.node,

  /**
   * The component to display in the panel content area.
   */
  panelContent: PropTypes.node,

  /**
   * The style of panel presentation. One of `overlay`, `squish`.
   */
  panelBehavior: PropTypes.oneOf(['overlay', 'squish']),

  /**
   * The position at which the panel will be displayed. This property honors the current direction setting. One of `start`, `end`.
   */
  panelPosition: PropTypes.oneOf(['start', 'end']),

  /**
   * The size at which the panel will be displayed. One of `small`, `large`.
   */
  panelSize: PropTypes.oneOf(['small', 'large']),

  /**
   * Whether or not, when open, the panel should be displayed with the full width of the SlidePanel.
   */
  isFullscreen: PropTypes.bool,

  /**
   * Whether or not the panel should be displayed.
   */
  isOpen: PropTypes.bool,

  /**
   * Whether or not the SlidePanel should be sized relative to its parent container.
   */
  fill: PropTypes.bool,

  /**
   * @private
   * Callback function to set the slide panel ref.
   */
  setSlidePanelRef: PropTypes.func,

  /**
   * @private
   * Intl object injected from injectIntl
   */
  intl: PropTypes.shape({ formatMessage: PropTypes.func, locale: PropTypes.string }),
};

const defaultProps = {
  panelBehavior: 'overlay',
  panelPosition: SlidePanelPositions.END,
  panelSize: 'small',
  replaceMainAriaDescribedBy: false,
};

class SlidePanel extends React.Component {
  constructor(props) {
    super(props);
    this.setPanelNode = this.setPanelNode.bind(this);
    this.mainNode = React.createRef();
    this.setLastClicked = this.setLastClicked.bind(this);
    this.setDisclosingNode = this.setDisclosingNode.bind(this);
    this.defaultMainAriaDescribedByID = 'detail-panel-warning';
    this.mainAriaDescribedByList = this.defaultMainAriaDescribedByID;

    if(SlidePanelUtils.isMac()){

      if (this.props.replaceMainAriaDescribedBy) {
        this.mainAriaDescribedByList = this.props.mainAriaDescribedBy;
      } else if (this.props.mainAriaDescribedBy) {
        this.mainAriaDescribedByList = this.mainAriaDescribedByList.concat(' ', this.props.mainAriaDescribedBy);
      }

    }
  }

  componentDidUpdate(prevProps) {
    if (this.props.isOpen && this.props.isOpen !== prevProps.isOpen) {
      // Save the disclosing node for returning focus when panel is closed
      this.setDisclosingNode(this.lastClicked);
      this.panelNode.focus();
      return;
    }

    if (!this.props.isOpen && this.props.isOpen !== prevProps.isOpen) {
      if (this.disclosingNode) {
        // Return focus to the disclosing element
        this.disclosingNode.setAttribute('aria-expanded', 'false');
        this.disclosingNode.focus();
        return;
      }
      // The disclosing element doesn't exist and return focus to the main div
      this.mainNode.current.focus();
    }
  }

  setPanelNode(node) {
    if (this.props.setSlidePanelRef) {
      this.props.setSlidePanelRef(node);
    }
    this.panelNode = node;
  }

  setLastClicked(event) {
    this.lastClicked = event.target;
  }

  setDisclosingNode(node) {
    if (node) {
      node.setAttribute('aria-expanded', 'true');
      this.disclosingNode = node;
    }
  }

  render() {
    const {
      intl,
      isFullscreen,
      isOpen,
      fill,
      mainAriaDescribedBy,
      mainAriaLabel,
      mainContent,
      panelAriaLabel,
      panelContent,
      panelBehavior,
      panelPosition,
      panelSize,
      replaceMainAriaDescribedBy,
      setSlidePanelRef,
      ...customProps
    } = this.props;

    const theme = this.context;

    const slidePanelClassNames = classNames(cx(
      'slide-panel',
      { 'is-open': isOpen },
      { 'is-fullscreen': isFullscreen },
      { fill },
      theme.className,
    ),
    customProps.className);

    const panelDiv = (
      <div
        className={cx(['panel'])}
        key="panel"
        tabIndex="-1"
        aria-label={panelAriaLabel || intl.formatMessage({ id: 'Terra.slidePanel.defaultPanelLabel' })}
        aria-hidden={!isOpen ? 'true' : 'false'}
        role="region"
        ref={this.setPanelNode}
      >
        <VisuallyHiddenText
          text={panelAriaLabel || intl.formatMessage({ id: 'Terra.slidePanel.defaultPanelLabel' })}
        />
        {panelContent}
      </div>
    );

    const mainDiv = (
      // eslint-disable-next-line jsx-a11y/no-static-element-interactions, jsx-a11y/no-noninteractive-element-interactions
      <div
        className={cx('main')}
        key="main"
        tabIndex="-1"
        aria-label={mainAriaLabel}
        aria-describedby={this.mainAriaDescribedByList}
        ref={this.mainNode}
        role="main"
        onClick={this.setLastClicked}
        onKeyUp={this.setLastClicked}
      >
        <VisuallyHiddenText
          tabIndex="-1"
          id={this.defaultMainAriaDescribedByID}
          text={intl.formatMessage({ id: 'Terra.slidePanel.discloseWarning' })}
        />
        {mainContent}
      </div>
    );

    const content = (panelPosition === SlidePanelPositions.START) ? (
      <React.Fragment>
        {panelDiv}
        {mainDiv}
      </React.Fragment>
    ) : (
      <React.Fragment>
        {mainDiv}
        {panelDiv}
      </React.Fragment>
    );

    return (
      <div
        {...customProps}
        className={slidePanelClassNames}
        data-slide-panel-panel-behavior={panelBehavior}
        data-slide-panel-panel-position={panelPosition}
        data-slide-panel-panel-size={panelSize}
      >
        {content}
      </div>
    );
  }
}

SlidePanel.propTypes = propTypes;
SlidePanel.defaultProps = defaultProps;
SlidePanel.contextType = ThemeContext;

export default injectIntl(SlidePanel);
export { SlidePanelPositions };
