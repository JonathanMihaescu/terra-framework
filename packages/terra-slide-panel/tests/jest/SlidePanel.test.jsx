import React from 'react';
import ThemeContextProvider from 'terra-theme-context/lib/ThemeContextProvider';
/* eslint-disable-next-line import/no-extraneous-dependencies */
import { shallowWithIntl, renderWithIntl, mountWithIntl } from 'terra-enzyme-intl';
import { IntlProvider } from 'react-intl';
import SlidePanel from '../../src/SlidePanel';
import translationsFile from '../../translations/en.json';

describe('When a SlidePanel is rendered', () => {
  it('should render a default SlidePanel with no props', () => {
    const slidePanel = <SlidePanel />;
    const wrapper = renderWithIntl(slidePanel);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render a SlidePanel with panelBehavior = squish', () => {
    const slidePanel = <SlidePanel panelBehavior="squish" />;
    const wrapper = renderWithIntl(slidePanel);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render a SlidePanel with panelBehavior = overlay', () => {
    const slidePanel = <SlidePanel panelBehavior="overlay" />;
    const wrapper = renderWithIntl(slidePanel);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render a SlidePanel with panelPosition = start', () => {
    const slidePanel = <SlidePanel panelPosition="start" />;
    const wrapper = renderWithIntl(slidePanel);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render a SlidePanel with panelPosition = end', () => {
    const slidePanel = <SlidePanel panelPosition="end" />;
    const wrapper = renderWithIntl(slidePanel);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render a SlidePanel with panelSize = small', () => {
    const slidePanel = <SlidePanel panelSize="small" />;
    const wrapper = renderWithIntl(slidePanel);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render a SlidePanel with panelSize = large', () => {
    const slidePanel = <SlidePanel panelSize="large" />;
    const wrapper = renderWithIntl(slidePanel);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render a SlidePanel with an open state', () => {
    const slidePanel = <SlidePanel isOpen />;
    const wrapper = renderWithIntl(slidePanel);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render a SlidePanel with a fullscreen state', () => {
    const slidePanel = <SlidePanel isFullscreen />;
    const wrapper = renderWithIntl(slidePanel);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render a SlidePanel with fill', () => {
    const slidePanel = <SlidePanel fill />;
    const wrapper = renderWithIntl(slidePanel);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render a SlidePanel with main content', () => {
    const slidePanel = (
      <SlidePanel
        mainContent={<p>Main Content</p>}
      />
    );
    const wrapper = renderWithIntl(slidePanel);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render a SlidePanel with panel content', () => {
    const slidePanel = (
      <SlidePanel
        panelContent={<p>Panel Content</p>}
      />
    );
    const wrapper = renderWithIntl(slidePanel);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render a SlidePanel with panelAriaLabel and mainAriaLabel specified', () => {
    const slidePanel = <SlidePanel panelAriaLabel="Panel content area" mainAriaLabel="Main content area" />;
    const wrapper = renderWithIntl(slidePanel);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render a SlidePanel with the given extra attributes', () => {
    const slidePanel = (
      <SlidePanel panelBehavior="overlay" id="my-slide-panel" className="pad-this-slide-panel" data-test-attr="ahoy thar" />
    );
    const wrapper = renderWithIntl(slidePanel);

    expect(wrapper).toMatchSnapshot();
  });

  it('should contain a Visually Hidden Text as the aria-describedby target', () => {
    const slidePanel = <SlidePanel />;
    const wrapper = shallowWithIntl(slidePanel).dive();

    const mainDiv = wrapper.find('#detail-panel-warning').prop('text');
    expect(mainDiv).toEqual('Terra.slidePanel.discloseWarning');
  });

  it('should set the text property of Visually Hidden Text from ./translations', () => {
    const slidePanel = <IntlProvider locale="en" messages={translationsFile}><SlidePanel /></IntlProvider>;
    const wrapper = shallowWithIntl(slidePanel).dive().dive();

    const mainDiv = wrapper.find('#detail-panel-warning').prop('text');
    expect(mainDiv).toEqual(translationsFile['Terra.slidePanel.discloseWarning']);
  });
});

describe('When custom props are used', () => {
  it('should override user provided attributes that aren\'t className', () => {
    const slidePanel = (
      <SlidePanel panelBehavior="overlay" className="pad-this-slide-panel" data-test-attr="ahoy thar" data-slide-panel-panel-behavior="i am not good with slide panel" />
    );
    const wrapper = renderWithIntl(slidePanel);

    expect(wrapper).toMatchSnapshot();
  });

  it('correctly applies the theme context className', () => {
    const tabs = mountWithIntl(
      <ThemeContextProvider theme={{ className: 'orion-fusion-theme' }}>
        <SlidePanel />
      </ThemeContextProvider>,
    );
    expect(tabs).toMatchSnapshot();
  });
});
