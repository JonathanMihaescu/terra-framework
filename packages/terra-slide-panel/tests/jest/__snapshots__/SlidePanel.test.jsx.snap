// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When a SlidePanel is rendered should render a default SlidePanel with no props 1`] = `
<SlidePanel
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": null,
      "now": [Function],
      "onError": [Function],
      "textComponent": "span",
      "timeZone": null,
    }
  }
  panelBehavior="overlay"
  panelPosition="end"
  panelSize="small"
/>
`;

exports[`When custom props are used correctly applies the theme context className 1`] = `
<ThemeContextProvider
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": null,
      "now": [Function],
      "onError": [Function],
      "textComponent": "span",
      "timeZone": null,
    }
  }
  theme={
    Object {
      "className": "orion-fusion-theme",
    }
  }
>
  <InjectIntl(SlidePanel)>
    <SlidePanel
      intl={
        Object {
          "defaultFormats": Object {},
          "defaultLocale": "en",
          "formatDate": [Function],
          "formatHTMLMessage": [Function],
          "formatMessage": [Function],
          "formatNumber": [Function],
          "formatPlural": [Function],
          "formatRelative": [Function],
          "formatTime": [Function],
          "formats": Object {},
          "formatters": Object {
            "getDateTimeFormat": [Function],
            "getMessageFormat": [Function],
            "getNumberFormat": [Function],
            "getPluralFormat": [Function],
            "getRelativeFormat": [Function],
          },
          "locale": "en",
          "messages": null,
          "now": [Function],
          "onError": [Function],
          "textComponent": "span",
          "timeZone": null,
        }
      }
      panelBehavior="overlay"
      panelPosition="end"
      panelSize="small"
    >
      <div
        className="slide-panel orion-fusion-theme"
        data-slide-panel-panel-behavior="overlay"
        data-slide-panel-panel-position="end"
        data-slide-panel-panel-size="small"
      >
        <div
          aria-describedby="detail-panel-warning"
          aria-hidden="false"
          className="main"
          key="main"
          onClick={[Function]}
          onKeyUp={[Function]}
          role="main"
          tabIndex="-1"
        >
          <VisuallyHiddenText
            id="detail-panel-warning"
            tabIndex="-1"
            text="Terra.slidePanel.discloseWarning"
          >
            <span
              className="visually-hidden-text"
              id="detail-panel-warning"
              tabIndex="-1"
            >
              Terra.slidePanel.discloseWarning
            </span>
          </VisuallyHiddenText>
        </div>
        <div
          aria-hidden="true"
          aria-label="Terra.slidePanel.defaultPanelLabel"
          className="panel"
          key="panel"
          role="region"
          tabIndex="-1"
        >
          <VisuallyHiddenText
            text="Terra.slidePanel.defaultPanelLabel"
          >
            <span
              className="visually-hidden-text"
            >
              Terra.slidePanel.defaultPanelLabel
            </span>
          </VisuallyHiddenText>
        </div>
      </div>
    </SlidePanel>
  </InjectIntl(SlidePanel)>
</ThemeContextProvider>
`;
